src/run_program.rs

```diff
@@ -133,6 +134,22 @@ fn traverse_path<T: Allocator>(
     Ok(Reduction(cost, arg_list))
 }
 
+fn augment_cost_errors<A: Clone>(
+    r: Result<Cost, EvalErr<A>>,
+    max_cost: &A,
+) -> Result<Cost, EvalErr<A>> {
+    if r.is_ok() {
+        return r;
+    }
+    let e = r.unwrap_err();
+    println!("error: {}", e.1);
+    if &e.1 != "cost exceeded" {
```

Comment:
this is not very pretty. However:

1. Our error messages are fundamentally based on strings (rather than error codes e.g.) so I don't see another way to distinguish error types.
2. The alternative, as far as I can tell, would be to pass this information down all the way everywhere this error can happen, which seems worse.

Reply:
another alternative (that I'm not sure I like any more) would be to store the max cost inside the allocator, to make it available to anyone creating this kind of error

Reply:
You might use `P` instead of `A` here to suggest `Ptr`. At first I thought `A` was for `Allocator`. Not a critical fix though.